### ISSUE TITLE
Make features, outputs, and labels more consistent

### ISSUE THREAD

**Issue by peastman** [deepchem/deepchem#1121]
I'm opening this to discuss issues I ran into while implementing `make_estimator()` in #1106.  We have a total of 19 model classes that extend TensorGraph, and 16 of them override `default_generator()`.  That's a big problem, because anything that overrides it will require extra work before it can be converted to an estimator.  I want to see if we can reduce that number.  Overriding `default_generator()` should be a rare thing, done only in exceptional circumstances.

Here are two principles I think we should try to follow.

1. It's the job of the featurizer to convert the input data to whatever format the model expects.

2. The outputs and the labels should have exactly the same shape and format.

Any time we violate either one of those principles, we'll get a model that can't be converted to an estimator without extra work (possibly a lot of extra work).  Sometimes it may be unavoidable, but not always.  Let me give some examples of models that violate them.

MultiTaskRegressor is a particularly gratuitous case.  It expects the labels to be a 2D tensor of shape (samples, tasks), but its output is a 3D tensor of shape (samples, tasks, 1).  The addition of that extra dimension  of size 1 is the only reason it needs to override `default_generator()`.

When we first designed TensorGraph, we came up with the principle that outputs should always be at least 3D with shape (samples, tasks, ...).  If there was only a single number per task, there would be a third dimension of size 1.  I've come to think that decision was a mistake.  Principle 2 above is much more important.  If we expect the labels to be 2D, the outputs should also be 2D.  For example, all the Tensorflow metrics expect that principle to be satisfied.  We're making our model more complicated just to create extra difficulties for ourselves.

MultiTaskClassifier is inconsistent in a related way.  It expects the labels to be 2D, with each value being a class index.  But the outputs are 3D, encoding probabilities for each class.  So it overrides `default_generator()` to convert the class indices into one-hot encoding, which is what the model's Label layer actually expects.

In both of these cases, we don't really need to override `default_generator()`.  We could just as easily have a layer in the graph add the extra dimension to the labels, or convert them to one-hot encoding.  Or for the classifier, we could use SparseSoftMaxCrossEntropy instead of SoftMaxCrossEntropy.  But that still wouldn't convert to an estimator correctly, because none of the Tensorflow metrics would work correctly.  They all expect the outputs and labels to be directly comparable.

**Comment by rbharath** [comment]
> I'm opening this to discuss issues I ran into while implementing make_estimator() in #1106. We have a total of 19 model classes that extend TensorGraph, and 16 of them override default_generator(). That's a big problem, because anything that overrides it will require extra work before it can be converted to an estimator. I want to see if we can reduce that number. Overriding default_generator() should be a rare thing, done only in exceptional circumstances.

I like the idea of decreasing the number of times we override default_generator(). Creating a new generator is one of the trickier parts of making a new TensorGraph, and reducing the need is good for us overall.

> Here are two principles I think we should try to follow.
>
> 1. It's the job of the featurizer to convert the input data to whatever format the model expects.
> 2. The outputs and the labels should have exactly the same shape and format.

Both of these sound like reasonable guidelines to follow. Are there any places where following them might make the API less friendly than it is right now?

**Comment by peastman** [comment]
> Are there any places where following them might make the API less friendly than it is right now?

I can only comment on the ones I'm familiar with, such as the ones I discussed above.  I don't know how most of the models work.  The authors of them will need to comment themselves.

> Creating a new generator is one of the trickier parts of making a new TensorGraph

It's now going to become a lot harder, because you'll also need to add the support for creating an estimator.  That will require doing all the same transformations you're already doing in `default_generator()`, except with Tensorflow instead of numpy.

**Comment by peastman** [comment]
Tensorflow is such a swamp of inconsistency and bad design, this whole thing is driving me crazy.  The output from MultiTaskClassifier is logits, not probabilities.  Of course it is.  We use `softmax_cross_entropy_with_logits()` to compute the loss.  They don't even offer a version that takes probabilities, so obviously we have to output logits.  Oh, but guess what?  `tf.metrics.accuracy()` expects the output to be probabilities, not logits.

Of course, this is another example for principle 2 above.  The outputs should have the same format as the labels, which means they should be probabilities, not logits.  But that means the loss will be computed based on an internal layer, not on the actual output values, which is certainly not intuitive.

**Comment by rbharath** [comment]
@peastman It might make sense to write up a blog-post on the DeepChem blog discussing these inconsistencies. We could communicate some of these issues to the TensorFlow team that way and they might be able to work on some upstream improvements

**Comment by rbharath** [comment]
I think these issues have been resolved by subsequent refactoring. Please re-open if there are still issues here.
